scripts/oracle-vuln-tns-poison.nse

```diff
@@ -0,0 +1,75 @@
+description = [[Detects the TNS Poison vulnerability.]]
+
+author      = "Alexandr Savca"
+license     = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories  = {"vuln", "safe"}
+
+---
+-- @usage
+-- nmap <ip> -p 1521-1523 --script oracle-vuln-tns-poison
+--
+-- @output
+-- PORT     STATE SERVICE
+-- 1521/tcp open  oracle
+-- | oracle-vuln-tns-poison:
+-- |   VULNERABLE:
+-- |   TNS Poison
+-- |     State: VULNERABLE
+-- |       The vulnerability affects the component called TNS Listener,
+-- |       which is the responsible of connections establishment. To
+-- |       exploit the vulnerability no privilege is needed, just network access to
+-- |       the TNS Listener.
+-- |
+-- |       The "feature" exploited is enabled by default in all
+-- |       Oracle versions starting with Oracle 8i and ending with Oracle 11g
+-- |       (without CPU-APR-2012).
+-- |
+-- |     Disclosure date: 2012-4-18
+-- |     References:
+-- |_      http://seclists.org/fulldisclosure/2012/Apr/204
+
+local stdnse = require "stdnse"
+local comm   = require "comm"
+local s_port = require "shortport"
+local vulns  = require "vulns"
+
+portrule = s_port.port_or_service({1521,1522,1523}, "oracle-tns")
+
+action = function(host, port)
+  local report = vulns.Report:new(SCRIPT_NAME, host)
+  local vuln   = {
+    title = 'TNS Poison',
+    state = vulns.STATE.NOT_VULN,
+    description = [[
+The vulnerability affects the component called TNS Listener,
+which is the responsible of connections establishment. To
+exploit the vulnerability no privilege is needed, just network access to
+the TNS Listener.
+
+The "feature" exploited is enabled by default in all
+Oracle versions starting with Oracle 8i and ending with Oracle 11g
+(without CPU-APR-2012).
+        ]],
+    references = {
+      'http://seclists.org/fulldisclosure/2012/Apr/204'
+    },
+    dates = {disclosure = {year = '2012', month = '4', day = '18'}}
+  }
+  local poison_probe = "\0\x68\0\0\x01\0\0\0\x016\x01,\0\0\x08\0\x7F\xFF\x7F\x08\0\0\0\x01\0 \0:\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\x004\xE6\0\0\0\x01\0\0\0\0\0\0\0\0(CONNECT_DATA=(COMMAND=service_register_NSGR))"
```

Comment:
Ideally, we'd dissect this to show what these bytes are. If you have a vulnerable service to test against, you could try using tns.lua functions to construct the packet.

Also, the original report says that `SERVICE_REGISTER_NSGR` is for Oracle 9i to 11g. It'd be great if we could use `SERVICE_REGISTER` for Oracle 8i if that was detected.